### PR TITLE
feat(plugin-bootstrap): Add multi-step workflow execution system with configurable strategy modes

### PR DESCRIPTION
**Pull request by tcm390**
# 🚀 Multi-Step Workflow Execution for ElizaOS Agents

## Overview
This PR introduces a sophisticated multi-step workflow execution system that enables ElizaOS agents to break down complex tasks into sequential, manageable steps. The system supports both single-shot and multi-step execution strategies, allowing agents to gather information through providers and execute actions in a controlled, iterative manner.

### 1. **Multi-Step Execution Engine**
- **Iterative Task Processing**: Agents can now execute tasks over multiple iterations, with configurable maximum iteration limits
- **Provider-Action Coordination**: Support for executing multiple providers followed by a single action per step
- **Progress Tracking**: Comprehensive tracing of each step's execution results and outcomes
- **Completion Detection**: Intelligent task completion detection with `isFinish` flag

### 2. **Enhanced Prompt Templates**
- **`multiStepDecisionTemplate`**: Guides agents through step-by-step decision making
- **`multiStepSummaryTemplate`**: Generates final summaries after task completion
- **Dynamic Context Building**: Leverages recent messages, action state, and provider results

### 3. **Flexible Execution Strategies**
- **Environment-Driven Toggle**: `USE_MULTI_STEP` environment variable controls execution mode
- **Fallback Support**: Graceful fallback to single-shot execution when multi-step is disabled
- **State Persistence**: Maintains execution state across iterations using runtime state cache

### 4. **Enhanced Provider System**
- **Comprehensive Provider Listing**: Shows both dynamic and static providers with descriptions
- **Provider Result Tracking**: Monitors provider execution success/failure for better debugging
- **Recent Message Context**: Enhanced context building with recent message analysis

### Core Changes
- **`packages/core/src/prompts.ts`**: Added multi-step decision and summary templates
- **`packages/plugin-bootstrap/src/index.ts`**: Implemented core multi-step execution logic
- **Test Infrastructure**: Comprehensive test coverage for multi-step workflows

### Architecture Improvements
- **State Management**: Enhanced state composition with action results tracking
- **Error Handling**: Robust error handling with detailed logging and fallback mechanisms
- **Performance**: Optimized execution flow with configurable iteration limits

### For Developers
- **Modular Design**: Clean separation between single-shot and multi-step execution
- **Extensible Framework**: Easy to add new execution strategies and templates
- **Comprehensive Testing**: Full test coverage ensures reliability

### For End Users
- **Complex Task Handling**: Agents can now tackle multi-step, information-gathering tasks
- **Better Context Awareness**: Enhanced provider usage leads to more informed responses
- **Progress Visibility**: Users can see step-by-step progress through callback mechanisms

### Test Coverage
- **Multi-step workflow execution tests**
- **Provider and action coordination tests**
- **Error handling and edge case scenarios**
- **State management and persistence tests**

### Test Utilities
- **Enhanced mock runtime with state cache support**
- **Comprehensive test data generation**
- **Integration test scenarios**

### Backward Compatibility
- **Single-shot execution remains the default**
- **Existing plugins and actions continue to work unchanged**
- **Gradual rollout through environment variable toggle**

### Configuration
```bash
# Enable multi-step execution
USE_MULTI_STEP=true

# Configure maximum iterations (default: 6)
MAX_MULTISTEP_ITERATIONS=8
```

### Standards Compliance
- **TypeScript strict mode compliance**
- **Comprehensive error handling**
- **Detailed logging and debugging support**
- **Clean code principles with clear separation of concerns**

### Performance Considerations
- **Configurable iteration limits prevent infinite loops**
- **Efficient state caching and retrieval**
- **Optimized provider and action execution**

### Prerequisites
- **Core package updates required**
- **Plugin-bootstrap package updates**
- **Environment variable configuration**

### Rollout Strategy
- **Feature flag controlled rollout**
- **Gradual adoption through configuration**
- **Monitoring and performance tracking**

---

## 📋 Checklist

- [x] **Core Functionality**: Multi-step execution engine implemented
- [x] **Prompt Templates**: Decision and summary templates added to core
- [x] **Provider Integration**: Enhanced provider system with result tracking
- [x] **State Management**: Improved state composition and persistence
- [x] **Error Handling**: Comprehensive error handling and logging
- [x] **Testing**: Full test coverage for all new functionality
- [x] **Documentation**: Clear code comments and inline documentation
- [x] **Performance**: Optimized execution flow with configurable limits
- [x] **Compatibility**: Backward compatible with existing functionality

## 🔗 Related Issues
- Implements multi-step workflow execution for complex agent tasks
- Enhances provider system with better result tracking
- Improves state management and context building
- Adds comprehensive testing infrastructure

---

This PR represents a significant enhancement to ElizaOS's agent capabilities, enabling more sophisticated task execution while maintaining backward compatibility and performance standards.

### CODE REVIEW

File: packages/plugin-bootstrap/src/providers/providers.ts [hunk 1]
```diff
@@ -20,42 +20,53 @@ export const providersProvider: Provider = {
...
-    const providerDescriptions = dynamicProviders.map((provider) => {
+    const dynamicDescriptions = dynamicProviders.map((provider) => {
       return `- **${provider.name}**: ${provider.description || 'No description available'}`;
     });
 
-    // Create the header text
-    const headerText =
-      '# Providers\n\nThese providers are available for the agent to select and use:';
+    const allDescriptions = allProviders.map((provider) => {
+      return `- **${provider.name}**: ${provider.description || 'No description available'}`;
```
Comment (odilitime) [inline packages/plugin-bootstrap/src/providers/providers.ts:34]: Maybe we shouldn’t include them and complain loudly in the logs if no description

File: packages/plugin-bootstrap/src/providers/providers.ts [hunk 2]
```diff
@@ -20,42 +20,53 @@ export const providersProvider: Provider = {
...
-    }
+    // Create the header text
+    const headerText = '# Providers\n\nThese providers are available for the agent to select and use:';
 
-    // Join all provider descriptions
-    const providersText = providerDescriptions.join('\n');
+    const dynamicSection =
+      dynamicDescriptions.length > 0
+        ? addHeader(headerText, dynamicDescriptions.join('\n'))
+        : addHeader(headerText, 'No dynamic providers are currently available.');
```
Comment (odilitime) [inline packages/plugin-bootstrap/src/providers/providers.ts:43]: Why waste the tokens, if there’s nothing there should be no text

File: packages/plugin-bootstrap/src/providers/providers.ts [hunk 3]
```diff
@@ -20,42 +20,53 @@ export const providersProvider: Provider = {
...
+      allDescriptions.join('\n')
+    );
 
-    // Also provide the data for potential programmatic use
     const data = {
       dynamicProviders: dynamicProviders.map((provider) => ({
         name: provider.name,
         description: provider.description || '',
       })),
+      allProviders: allProviders.map((provider) => ({
```
Comment (odilitime) [inline packages/plugin-bootstrap/src/providers/providers.ts:55]: Why include these?

File: packages/core/src/prompts.ts [hunk 4]
```diff
@@ -161,3 +161,75 @@ Respond using XML format like this:
...
+{{actionResults}}
+
+<output>
+<response>
+  <thought>
+    Clearly explain your reasoning for the next step and how it helps resolve the user's request. 
+    If calling an action or provider, mention the relevant parameters you are passing.  
+    Example: "To proceed, I need to call ACTION_NAME with parameters { 'key': 'value' } to gather the necessary data."
+  </thought>
+  <nextStepType>action | provider | finish</nextStepType>
```
Comment (odilitime) [inline packages/core/src/prompts.ts]: Oof, really really slow (and costly with inference)

File: packages/plugin-bootstrap/src/providers/providers.ts (hunk 1 above)
Comment (tcm390) [reply packages/plugin-bootstrap/src/providers/providers.ts:34]: yea should remove it 🥲

File: packages/core/src/prompts.ts [hunk 5]
```diff
@@ -161,3 +161,75 @@ Respond using XML format like this:
...
+These are the actions or data provider calls that have already been used in this run. Use this to avoid redundancy and guide your next move.
+
+{{actionResults}}
+
+<output>
+<response>
+  <thought>
+    Clearly explain your reasoning for the next step and how it helps resolve the user's request. 
+    If calling an action or provider, mention the relevant parameters you are passing.  
+    Example: "To proceed, I need to call ACTION_NAME with parameters { 'key': 'value' } to gather the necessary data."
```
Comment (odilitime) [inline packages/core/src/prompts.ts]: Shouldn’t we break out the parameters to their own field?

File: packages/plugin-bootstrap/src/index.ts [hunk 6]
```diff
@@ -378,8 +379,6 @@ const messageReceivedHandler = async ({
       startTime,
       status: 'started',
       source: 'messageHandler',
-      // this shouldn't be a standard
-      // but we need to expose content somewhere
```
Comment (odilitime) [inline packages/plugin-bootstrap/src/index.ts]: Why are we removing documentation?

File: packages/plugin-bootstrap/src/providers/providers.ts (hunk 3 above)
Comment (tcm390) [reply packages/plugin-bootstrap/src/providers/providers.ts:55]: So now, in the decision step, we're no longer including static providers in the prompt. Instead, we let the agent decide whether it needs a provider, and only then call it. Essentially, this makes all providers dynamic. Not sure if this is the ideal approach...

**Review by standujar — APPROVED** [review]
